src/proxy/http/metrics/handle_time.rs

```diff
@@ -0,0 +1,235 @@
+use metrics::{latency, FmtLabels, FmtMetric, Histogram};
+use proxy::http::insert;
+use std::{
+    fmt,
+    sync::{
+        atomic::{self, AtomicUsize, Ordering},
+        Arc, Mutex, RwLock,
+    },
+    time::Instant,
+};
+
+/// A single handle time histogram.
+///
+/// Higher-level code will use this to represent a single set of labels for
+/// handle-time metrics.
+#[derive(Debug, Clone)]
+pub struct Scope(Arc<Shared>);
+
+/// A layer that inserts a `Tracker` into each request passing through it.
+#[derive(Debug, Clone)]
+pub struct InsertTracker(Arc<Shared>);
+
+/// A request extension that, when dropped, records the time elapsed since it
+/// was created.
+#[derive(Debug)]
+pub struct Tracker {
+    shared: Arc<Shared>,
+    idx: usize,
+    t0: Instant,
+}
+
+#[derive(Debug)]
+struct Shared {
+    // NOTE: this is inside a `Mutex` since recording a latency requires a mutable
+    // reference to the histogram. In the future, we could consider making the
+    // histogram counters `AtomicU64, so that the histogram could be updated
+    // with an immutable reference. Then, the mutex could be removed.
+    histogram: Mutex<Histogram<latency::Us>>,
+    // Stores the state of currently active `Tracker`s.
+    recorders: RwLock<Vec<Recorder>>,
+    // The next free index into `recorders`
+    next_recorder: AtomicUsize,
+}
+
+#[derive(Debug)]
+struct Recorder {
```

Comment:
You mentioned this a little in the PR comment, but is the only purpose of `ref_count` in `Recorder` that fact that there can multiple references to it in the case of retries? I found this a little confusing to understand until coming back to that comment.

Also are there cases when next is not just `i + 1`?

Reply:
> You mentioned this a little in the PR comment, but is the only purpose of `ref_count` in `Recorder` that fact that there can multiple references to it in the case of retries? I found this a little confusing to understand until coming back to that comment.

Perhaps the name could be made clearer...would `trackers` or `clones` be better?

> Also are there cases when next is not just `i + 1`?

Yes, when all the `Tracker`s for a given index are dropped, it's reused so we don't have to allocate more storage for the next request. Here's the relevant code:
https://github.com/linkerd/linkerd2-proxy/blob/e8f5d136d4362e76ea678a1d89364e354b85ed8b/src/proxy/http/metrics/handle_time.rs#L140-L174

Reply:
Oh okay I see [here](https://github.com/linkerd/linkerd2-proxy/blob/0231f60979e9d988635c1f0f90b2a72a677681c3/src/proxy/http/metrics/handle_time.rs#L225-L226) that the swap makes sure that `recorders.len()` always should return the first free index.